tokio-sync/src/lock.rs

```diff
@@ -0,0 +1,202 @@
+//!         }
+//!     }
+//! }
+//! ```
+//!
+//!   [`Lock`]: struct.Lock.html
+//!   [`LockGuard`]: struct.LockGuard.html
+
+use futures::{Async, Future, Poll, Sink, StartSend, Stream};
+use semaphore;
+use std::cell::UnsafeCell;
+use std::fmt;
+use std::ops::{Deref, DerefMut};
+use std::sync::Arc;
+
+/// An asynchronous mutual exclusion primitive useful for protecting shared data
+///
+/// Each mutex has a type parameter (`T`) which represents the data that it is protecting. The data
+/// can only be accessed through the RAII guards returned from `poll_lock`, which guarantees that
+/// the data is only ever accessed when the mutex is locked.
+#[derive(Debug)]
+pub struct Lock<T> {
+    inner: Arc<State<T>>,
+    permit: semaphore::Permit,
+}
+
+/// A handle to a held `Lock`.
+///
+/// As long as you have this guard, you have exclusive access to the underlying `T`. The guard
+/// internally keeps a reference-couned pointer to the original `Lock`, so even if the lock goes
+/// away, the guard remains valid.
+///
+/// The lock is automatically released whenever the guard is dropped, at which point `poll_lock`
+/// will succeed yet again.
+#[derive(Debug)]
+pub struct LockGuard<T>(Lock<T>);
+
+// As long as T: Send, it's fine to send Lock<T> to other threads.
+// If T was not Send, sending a Lock<T> would be bad, since you can access T through Lock<T>.
+unsafe impl<T> Send for Lock<T> where T: Send {}
+unsafe impl<T> Sync for LockGuard<T> where T: Send + Sync {}
+
+#[derive(Debug)]
+struct State<T> {
+    c: UnsafeCell<T>,
+    s: semaphore::Semaphore,
+}
+
+impl<T> Lock<T> {
+    /// Try to acquire the lock.
+    ///
+    /// If the lock is already held, the current task is notified when it is released.
+    pub fn poll_lock(&mut self) -> Async<LockGuard<T>> {
+        if let Async::NotReady = self.permit.poll_acquire(&self.inner.s).unwrap_or_else(|_| {
+            // The semaphore was closed. but, we never explicitly close it, and we have a
+            // handle to it through the Arc, which means that this can never happen.
+            unreachable!()
+        }) {
+            return Async::NotReady;
+        }
+
+        // We want to move the acquired permit into the guard,
+        // and leave an unacquired one in self.
+        let acquired = Self {
+            inner: self.inner.clone(),
+            permit: ::std::mem::replace(&mut self.permit, semaphore::Permit::new()),
+        };
+        Async::Ready(LockGuard(acquired))
+    }
+}
+
+impl<T> Drop for LockGuard<T> {
+    fn drop(&mut self) {
+        if self.0.permit.is_acquired() {
+            self.0.permit.release(&self.0.inner.s);
+        } else if ::std::thread::panicking() {
+            // A guard _should_ always hold its permit, but if the thread is already panicking,
+            // we don't want to generate a panic-while-panicing, since that's just unhelpful!
+        } else {
+            unreachable!("Permit not help when LockGuard was dropped")
```

Comment:
I think this is a typo:
```suggestion
            unreachable!("Permit not held when LockGuard was dropped")
```

Reply:
Yup! Fixed.